spinalcordtoolbox/deepseg/inference.py

```diff
@@ -83,3 +94,158 @@ def segment_and_average_volumes(model_paths, input_filenames, options):
+        inference = segment_monai
+    else:
+        assert model_type == "nnunet"
+        create_net = ds_nnunet.create_nnunet_from_plans
+        inference = segment_nnunet
+
+    # equivalent to `with torch.no_grad()`
+    torch.set_grad_enabled(False)
+
+    # load model from checkpoint
+    net = create_net(path_model)
+
+    im_lst, target_lst = [], []
+    for fname_in in input_filenames:
+        # model may be multiclass, so the `inference` func should output a list of fnames and targets
+        fnames_out, targets = inference(path_img=fname_in, tmpdir=tmp_create(basename="sct_deepseg"), predictor=net)
+        for fname_out, target in zip(fnames_out, targets):
+            im_out = Image(fname_out)
+            if threshold is not None:
+                im_out.data = binarize(im_out.data, threshold)
+            im_lst.append(im_out)
+            target_lst.append(target)
+
+    return im_lst, target_lst
+
+
+def segment_monai(path_img, tmpdir, predictor):
+    """
+    Script to run inference on a MONAI-based model for contrast-agnostic soft segmentation of the spinal cord.
+
+    Author: Naga Karthik
+    Original script: https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/e65478099d026f865b7f1d7d0082e6e9a507a744/monai/run_inference_single_image.py
+    """
+    # Copy the file to the temporary directory using shutil.copyfile
+    path_img_tmp = os.path.join(tmpdir, os.path.basename(path_img))
+    shutil.copyfile(path_img, path_img_tmp)
+    logger.info(f'Copied {path_img} to {path_img_tmp}')
+
+    # define inference patch size and center crop size
+    crop_size = (64, 192, -1)
+    inference_roi_size = (64, 192, 320)
+
+    # define the dataset and dataloader
+    test_loader, test_post_pred = ds_monai.prepare_data(path_img_tmp, crop_size=crop_size)
+    batch = next(iter(test_loader))
+
+    # Run MONAI prediction
+    print('Starting inference...')
+    start = time.time()
+
+    # run inference
+    test_input = batch["image"].to(torch.device("cpu"))
+    batch["pred"] = sliding_window_inference(test_input, inference_roi_size, mode="gaussian",
+                                             sw_batch_size=4, predictor=predictor, overlap=0.5, progress=False)
+    pred = ds_monai.postprocessing(batch, test_post_pred)
+
+    end = time.time()
+    print('Inference done.')
+    total_time = end - start
+    print(f'Total inference time: {int(total_time // 60)} minute(s) {int(round(total_time % 60))} seconds')
+
+    # this takes about 0.25s on average on a CPU
+    # image saver class
+    _, fname, ext = extract_fname(path_img)
+    postfix = "seg"
+    target = f"_{postfix}"
+    pred_saver = SaveImage(
+        output_dir=tmpdir, output_postfix=postfix, output_ext=ext,
+        separate_folder=False, print_log=False)
+    # save the prediction
+    fname_out = os.path.join(tmpdir, f"{fname}_{postfix}{ext}")
+    logger.info(f"Saving results to: {fname_out}")
+    pred_saver(pred)
```

Comment:
I looked at `SaveImage`, and noticed that [it has a `savepath_in_metadict` option](https://docs.monai.io/en/stable/transforms.html#saveimage), which would be arguably safer to use than recomputing `fname_out` ourselves, in case they update their naming scheme. But, looking at [the source code](https://docs.monai.io/en/stable/_modules/monai/transforms/io/array.html#SaveImage), there's an extra condition that `meta_data is not None`, and I'm not sure yet:
* how to ensure that `meta_data` is not `None`; and
* how to access the right `meta_data` dict.

Given that their code is a little convoluted, maybe it's actually safer to do the simple thing, that is, computing `fname_out` ourselves as this code already does?